### PR TITLE
Fix #6750: Avatar added Gravatar support

### LINKED ISSUE #6750: Avatar: Port Gravatar from Extensions
**Description**
PrimeFaces Extensions has a Gravatar component: https://www.primefaces.org/showcase-ext/sections/gravatar/basic.jsf

See: https://github.com/primefaces-extensions/primefaces-extensions/issues/258

**Describe the solution you would like**
Port the functionality to the new PF Avatar component to allow Gravatars.

**Example XHTML**
```xml
<h:form id="frmTest">
  <p:avatar gravatar="pfextensions@gmail.com" gravatarConfig="d=mp" />
</h:form>
```

image

### PR DESCRIPTION
**Pull request by melloware**
Ported from Extensions

### CODE REVIEW

File: docs/9_0/components/avatar.md [hunk 1]
```diff
@@ -27,6 +27,8 @@ Avatar represents people using icons, labels and images.
 | shape | square | String | Shape of the element, valid options are "square" and "circle".
 | style | null | String | Style of the avatar.
 | styleClass | null | String | StyleClass of the avatar.
+| gravatar | null | String | SGravatar e-mail address for the profile to be displayed.
```
Comment (christophs78) [inline docs/9_0/components/avatar.md]: typo SGravatar

Comment (melloware) [reply docs/9_0/components/avatar.md]: Thanks fixed!

**Review by buunsal — APPROVED** [review]

**Comment by melloware** [comment]
This has been rebased with Mert's latest changes

File: src/main/java/org/primefaces/component/avatar/AvatarRenderer.java [hunk 2]
```diff
@@ -98,4 +110,48 @@ public void encodeChildren(FacesContext context, UIComponent component) throws I
...
+     * @see https://en.gravatar.com/site/implement/images/
+     * @param avatar the Avatar to create a Gravatar for
+     * @return the URL to retrieve the Gravatar image
+     */
+    protected String generateGravatar(Avatar avatar) {
+        String email = avatar.getGravatar();
+        String config = avatar.getGravatarConfig();
+        String url = null;
+        try {
+            final StringBuilder sb = new StringBuilder(1024);
```
Comment (tandraschko) [inline src/main/java/org/primefaces/component/avatar/AvatarRenderer.java]: - remove final
- use SharedStringBuilder

**Comment by melloware** [comment]
Removed all `final` and switched to SharedStringBuilder and rebased from master.

File: src/main/java/org/primefaces/component/avatar/AvatarRenderer.java [hunk 3]
```diff
@@ -92,4 +106,49 @@ public void encodeChildren(FacesContext context, UIComponent component) throws I
...
+    }
+
+    /**
+     * Converts email address into a hash code for Gravatar API.
+     *
+     * @param email the email to encode
+     * @return a MD5 hashcode of the email address
+     * @throws NoSuchAlgorithmException if hash can't be encoded
+     */
+    protected String generateMailHash(String email) throws NoSuchAlgorithmException {
```
Comment (tandraschko) [inline src/main/java/org/primefaces/component/avatar/AvatarRenderer.java]: why not pass the existing StringBuilder here?

Comment (melloware) [reply src/main/java/org/primefaces/component/avatar/AvatarRenderer.java]: Great idea. Done!

**Comment by tandraschko** [comment]
@melloware please merge it after you fixed the conflicts :D

**Pull request merged**